CSS content update: learn to support older browsers (#31473)
mdn/content: files/en-us/learn/css/css_layout/supporting_older_browsers/index.md

## Support doesn't mean "looks the same"

A website can't possibly look the same in all browsers. Some of your users will be viewing the site on a phone and others on a large desktop screen. Similarly, some of your users will have an old browser version, and others the latest browser. Some of your users might be hearing your content read out to them by a screen reader, or are zoomed in on the page to be able to read it. Supporting everyone means serving a version of your content that is designed defensively, so that it will look great on modern browsers, but will still be usable at a basic level for all users no matter how they are accessing your content.

A basic level of support comes from structuring your content well so that the normal flow of your page makes sense. A user with a very limited data plan might not load images, fonts, or even your CSS, but the content should still be presented in a way that makes reading easy. A well-structured HTML document should always be your starting point. _If you remove your stylesheet, does your content make sense?_

> It doesn't make commercial sense to pour time into trying to give everyon identical experiences as user environments vary and is not something you can controll. It is helpful to render a plain, CSS-less view of your site to ensure a fallback experience of your site is accessible. This fallback may never be viewed by people using very old or limited browsers; but may be viewed by your main target audience — users of modern browsers — when their browser or Internet connection temporarily fails. It is better to focus on what you can control: it is better to spend the time to make the site [accessible](/en-US/docs/Web/Accessibility), thus serving more users. There is a middle ground between a plain HTML page and all the bells and whistles, and CSS makes creating fallbacks pretty straightforward.

Reviewer: ```suggestion
It doesn't make commercial sense to pour time into trying to give everyon identical experiences as user environments vary and is not something you can control. It is helpful to render a plain, CSS-less view of your site to ensure a fallback experience of your site is accessible. This fallback may never be viewed by people using very old or limited browsers; but may be viewed by your main target audience — users of modern browsers — when their browser or Internet connection temporarily fails. It is better to focus on what you can control: it is better to spend the time to make the site [accessible](/en-US/docs/Web/Accessibility), thus serving more users. There is a middle ground between a plain HTML page and all the bells and whistles, and CSS makes creating fallbacks pretty straightforward.
```

CSS specifications contain information that explains what the browser does when two similar features, such as layout methods, are applied to the same item. For example, there is a definition for what happens if a floated item is also a grid item using CSS grid container. There is also a definition for what happens when an element has both a `margin-top` and a `margin-block-start` set.

These techniques were demonstrated in the [legacy techniques layout tutorial](/en-US/docs/Learn/CSS/CSS_layout/Legacy_Layout_Methods). Back when grid wasn't fully supported, the browser ignored the grid styles. This tutorial is now a good demonstration of how newer features, in this case grid layout, are designed and defined to take precedence over older features, floats in this case, when their effects would otherwise conflict.

### Fallback methods

> As noted above, a browser will discard any CSS it doesn't understand. Check out the [CSS error handling guide](/en-US/docs/Web/CSS/CSS_syntax/Error_handling). This enables using new CSS properties and new property values before they are universally supported. New CSS features that aren't understood are discarded as invalid. This enabled the use of [vendor-prefixing](/en-US/docs/Web/CSS/CSS_syntax/Error_handling#vendor_prefixes)

Reviewer: ```suggestion
As noted above, a browser will discard any CSS it doesn't understand. Check out the [CSS error handling guide](/en-US/docs/Web/CSS/CSS_syntax/Error_handling). This enables using new CSS properties and new property values before they are universally supported. New CSS features that aren't understood are discarded as invalid. This enables the use of [vendor-prefixing](/en-US/docs/Web/CSS/CSS_syntax/Error_handling#vendor_prefixes)
```

## What is the browser landscape for your site?

Every website is different in terms of its target audience. Before deciding on the approach to take, find out the number of visitors coming to your site using older browsers. This is straightforward if you have an existing website you are adding to or replacing, as you probably have analytics available that can tell you the technology people are using. If you have no analytics or this is a brand new site, then there are sites such as [Statcounter](https://gs.statcounter.com/) that can provide statistics filtered by location.

> You should also consider the type of devices and the way people use your site. For example, you may expect a higher-than-average number of mobile devices. Accessibility and people using assistive technology should always be considered, but for some sites that may be even more critical. In my experience, developers are often very worried about the experience of 1% of users using the particular browser and operating system combination as their company's CEO, while not considering the far greater number who have accessibility needs.

Author: that actually was a huge problem in the teens: all devs used macs and all CEOs were running IE. likely not an issue anymore (but glad i don't work for a CEO like that (anymore))

## Support doesn't mean "looks the same"

A website can't possibly look the same in all browsers. Some of your users will be viewing the site on a phone and others on a large desktop screen. Similarly, some of your users will have an old browser version, and others the latest browser. Some of your users might be hearing your content read out to them by a screen reader, or are zoomed in on the page to be able to read it. Supporting everyone means serving a version of your content that is designed defensively, so that it will look great on modern browsers, but will still be usable at a basic level for all users no matter how they are accessing your content.

A basic level of support comes from structuring your content well so that the normal flow of your page makes sense. A user with a very limited data plan might not load images, fonts, or even your CSS, but the content should still be presented in a way that makes reading easy. A well-structured HTML document should always be your starting point. _If you remove your stylesheet, does your content make sense?_

> It doesn't make commercial sense to pour time into trying to give everyon identical experiences as user environments vary and is not something you can control. It is helpful to render a plain, CSS-less view of your site to ensure a fallback experience of your site is accessible. This fallback may never be viewed by people using very old or limited browsers; but may be viewed by your main target audience — users of modern browsers — when their browser or Internet connection temporarily fails. It is better to focus on what you can control: it is better to spend the time to make the site [accessible](/en-US/docs/Web/Accessibility), thus serving more users. There is a middle ground between a plain HTML page and all the bells and whistles, and CSS makes creating fallbacks pretty straightforward.

Reviewer 2: - Replaced some phrases to keep it simple for non-native English speakers
- The last sentence is pretty abrupt and does not tie with the idea of the paragraph
```suggestion
It doesn't make commercial sense to spend time trying to give everyone an identical experience of your website. This is because user environments can vary widely and are beyond your control. It is helpful to test a plain, CSS-less view of your site to ensure that the fallback experience of your site is accessible. This fallback may never be viewed by people using very old or limited browsers, but may be viewed by your main target audience — users of modern browsers — when their browser or Internet connection fails temporarily. It is better to focus on what you can control: it is better to spend the time to make the site [accessible](/en-US/docs/Web/Accessibility), thus serving more users. There is a balance between a plain HTML page and a fully featured website. CSS simplifies creating these fallbacks.
```

> It doesn't make commercial sense to spend time trying to give everyone an identical experience of your website. This is because user environments can vary widely and are beyond your control. It is helpful to test a plain, CSS-less view of your site to ensure that the fallback experience of your site is accessible. This fallback may never be viewed by people using very old or limited browsers, but may be viewed by your main target audience — users of modern browsers — when their browser or Internet connection fails temporarily. It is better to focus on what you can control: it is better to spend the time to make the site [accessible](/en-US/docs/Web/Accessibility), thus serving more users. There is a balance between a plain HTML page and a fully featured website. CSS simplifies creating these fallbacks.

Reviewer 2: Moving the last two sentences around to make them flow with the rest of the para
```suggestion
It doesn't make commercial sense to spend time trying to give everyone an identical experience of your website. This is because user environments can vary widely and are beyond your control. There is a balance you need to strike between a plain HTML page and a fully featured website. It is helpful to test a plain, CSS-less view of your site to ensure that the fallback experience of your site is accessible. This fallback may never be viewed by people using very old or limited browsers, but may be viewed by your main target audience — users of modern browsers — when their browser or Internet connection fails temporarily. CSS simplifies creating these fallbacks. Therefore, it is better to focus on what you can control, that is, spend the time to make your site [accessible](/en-US/docs/Web/Accessibility), thereby serving more users.
```

## Creating fallbacks in CSS

> Browsers ignore CSS that they don't understand. When a browser doesn't recognize a new feature, it [discards the declaration as invalid](/en-US/docs/Web/CSS/CSS_syntax/Error_handling#css_parser_errors) without throwing an error. This allows both old and new values to coexist in the same ruleset. Just make sure to declare the old value before the new value so that, when supported, the new value overwrites the fallback value.

Reviewer 2: ```suggestion
Browsers ignore the CSS properties and values they don't support. When a browser doesn't recognize a new feature, it [discards the declaration as invalid](/en-US/docs/Web/CSS/CSS_syntax/Error_handling#css_parser_errors) without throwing an error. This allows both old and new values to coexist in the same ruleset. Just make sure to declare the old value before the new value so that, when supported, the new value overwrites the old value (the fallback).
```

> For example, most browsers support the two-value {{cssxref("display")}} syntax. If a browser doesn't, it will heed the legacy single value syntax.

Reviewer 2: ```suggestion
For example, most browsers support the two-value syntax of the {{cssxref("display")}} property. If a browser doesn't, it will use the older, single-value syntax.
```

> When a browser doesn't recognize a new feature, it [discards the declaration as invalid](/en-US/docs/Web/CSS/CSS_syntax/Error_handling#css_parser_errors) without throwing an error. Because browsers discard CSS properties and values they don't support, both old and new values can coexist in the same ruleset. Just make sure to declare the old value before the new value so that, when supported, the new value overwrites the old value (the fallback).

Reviewer 2: Perhaps the link can be moved to a different phrase
```suggestion
When a browser doesn't recognize a new feature, it discards the declaration as invalid [without throwing an error](/en-US/docs/Web/CSS/CSS_syntax/Error_handling#css_parser_errors). Because browsers discard CSS properties and values they don't support, both old and new values can coexist in the same ruleset. Just make sure to declare the old value before the new value so that, when supported, the new value overwrites the old value (the fallback).
```

```css
.container {
  display: inline-flex;
  display: inline flex;
}
```

> CSS specifications contain information that explains what the browser does when two similar features, such as layout methods, are applied to the same item. For example, there is a definition for what happens if a floated item is also a grid item using CSS grid container. There is also a definition for what happens when an element has both a {{cssxref("margin-top")}} and a {{cssxref("margin-block-start")}} set.

Reviewer 2: ```suggestion
CSS specifications contain information that explain what the browser does when two similar features, such as layout methods, are applied to the same item. For example, they define what happens if an item is floated and is also a grid item and part of a CSS grid container. There is also a definition for what happens when an element has both {{cssxref("margin-top")}} and {{cssxref("margin-block-start")}} properties set.
```

> These techniques were demonstrated in the [legacy techniques layout tutorial](/en-US/docs/Learn/CSS/CSS_layout/Legacy_Layout_Methods). Back when grid wasn't fully supported, the browser ignored the grid styles. This tutorial is now a good demonstration of how newer features, in this case grid layout, are designed and defined to take precedence over older features, floats in this case, when their effects would have otherwise conflicted.

Reviewer 2: ```suggestion
These techniques are demonstrated in the [legacy techniques layout tutorial](/en-US/docs/Learn/CSS/CSS_layout/Legacy_Layout_Methods). Back when grid wasn't fully supported, the browser ignored the grid styles. This tutorial is now a good demonstration of how newer features, in this case grid layout, are designed and defined to take precedence over older features, floats in this case, when their effects would have otherwise conflicted.
```

Reviewer 2: I guess we removed this para?

> Similarly, this [error-handling](/en-US/docs/Web/CSS/CSS_syntax/Error_handling#vendor_prefixes) ensures old CSS code bases continue to work even if legacy {{glossary("vendor-prefix", "vendor-prefixed")}} features are no longer supported. While vendor prefixing is no longer commonly used, if you must include a vendor-prefixed property or value, make sure to declare the prefixed value before the standard value so that, when supported, the new value overwrites the fallback value.

Reviewer 2: ```suggestion
Similarly, this [error-handling](/en-US/docs/Web/CSS/CSS_syntax/Error_handling#vendor_prefixes) ensures old CSS code bases continue to work even if legacy {{glossary("Vendor_Prefix", "vendor-prefixed")}} features are no longer supported. While vendor prefixing is no longer commonly used, if you must include a vendor-prefixed property or value, make sure to declare the prefixed value before the standard value so that, when supported, the new value overwrites the fallback value.
```

### Fallback methods

> As noted above, a browser will discard any CSS it doesn't understand. Check out the [CSS error handling guide](/en-US/docs/Web/CSS/CSS_syntax/Error_handling). This enables using new CSS properties and new property values before they are universally supported. New CSS features that aren't understood are discarded as invalid. This enabled the use of [vendor-prefixing](/en-US/docs/Web/CSS/CSS_syntax/Error_handling#vendor_prefixes), and enables legacy CSS to continue to work in spite of legacy vendor-prefixes that are no longer supported.

Reviewer 2: - The link to the guide is abrupt and should probably be moved to the end of the para. The sentence after the link to the guide should follow the first sentence
```suggestion
As noted above, a browser will discard any CSS it doesn't support. Check out the [CSS error handling guide](/en-US/docs/Web/CSS/CSS_syntax/Error_handling). This browser feature enables using new CSS properties and new property values before they are universally supported. New CSS features that aren't supported are discarded as invalid. This enabled the use of [vendor-prefixing](/en-US/docs/Web/CSS/CSS_syntax/Error_handling#vendor_prefixes), and also enables legacy CSS to continue to work in spite of legacy vendor-prefixes that are no longer supported.
```

> While vendor prefixing is no longer commonly used, if you must include a vendor prefixed property or value, make sure to declare the old value before the new value so that, when supported, the new value overwrites the fallback value. If using vendor prefixed [pseudo-elements](/en-US/docs/Web/CSS/Pseudo-elements) or [pseudo-classes](/en-US/docs/Web/CSS/Pseudo-classes), include the prefixed values within a [forgiving selector list](/en-US/docs/Web/CSS/Selector_list#forgiving_selector_list), using {{cssxref(":is", ":is()")}} or {{cssxref(":where", ":where()")}} so the entire selector block doesn't get [invalidated and ignored](/en-US/docs/Web/CSS/Selector_list#invalid_selector_list).

Reviewer 2: - Should probably link to https://developer.mozilla.org/en-US/docs/Web/CSS/Mozilla_Extensions#pseudo-elements_and_pseudo-classes
```suggestion
While vendor prefixing is no longer commonly used, if you must include a vendor prefixed property or value, make sure to declare the old value before the new value so that, when supported, the new value overwrites the fallback value. If using vendor-prefixed [pseudo-elements](/en-US/docs/Web/CSS/Pseudo-elements) or [pseudo-classes](/en-US/docs/Web/CSS/Pseudo-classes), include the prefixed values within a [forgiving selector list](/en-US/docs/Web/CSS/Selector_list#forgiving_selector_list) by using {{cssxref(":is", ":is()")}} or {{cssxref(":where", ":where()")}} so the entire selector block doesn't get [invalidated and ignored](/en-US/docs/Web/CSS/Selector_list#invalid_selector_list).
```

```css
:is(:-prefix-mistake, :invalid-error),
.valid {
  font-family: sans-serif;
}
:-prefix-mistake,
:invalid-error,
.valid {
  color: red;
}
```

> In the above example, the `.valid` content will be sans-serif but not red.

Reviewer 2: ```suggestion
In the above example, the `.valid` content will be `sans-serif` but not `red`.
```

## Feature queries

> Feature queries allow you to test whether a browser supports any particular CSS feature. This means that you can write some CSS for browsers that don't support a certain feature, then check to see if the browser has support and if so throw in your fancy new features.

Reviewer 2: ```suggestion
Feature queries allow you to test whether a browser supports a particular CSS feature. This means that you can write some CSS for browsers that don't support a certain feature, then check to see if the browser has support and if so, throw in your fancy new features.
```

  <div class="item">Item One</div>
  <div class="item">Item Two</div>
  <div class="item">Item Three</div>
  <div class="item">Item Four</div>
  <div class="item">Item Five</div>
  <div class="item">Item Six</div>
</div>
```

{{ EmbedLiveSample('Feature_queries', '100%', '200') }}

> Feature queries are supported in all modern browsers. Write your CSS for fully supported features first, outside of any feature queries. Once your site is usable and accessible to all users, add new features within feature query blocks. Browsers that support the feature queried can then render the newer CSS inside the feature query block. Use the approach of writing well supported CSS first, then enhancing features based on support.

Reviewer 2: ```suggestion
Feature queries are supported in all modern browsers. Write your CSS for fully supported features first, outside of any feature queries. Once your site is usable and accessible to all users, add new features within feature query blocks. Browsers that support the feature queried can then render the newer CSS inside the feature query block. Use the approach of writing well-supported CSS first, then enhancing features based on support.
```

## Testing older browsers

One way is to use an online testing tool such as Sauce Labs, as detailed in the [Cross browser testing](/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing) module.

## Summary

You now have the knowledge to provide fallback CSS for older browsers and confidently test for new features. You should now feel confident making use of any new techniques that might come along.

Now that you have worked through our articles on CSS layout, it's time to test your comprehension with our assessment for the module: [Fundamental layout comprehension](/en-US/docs/Learn/CSS/CSS_layout/Fundamental_Layout_Comprehension).

## See also

- [`@supports`](/en-US/docs/Web/CSS/@supports) at-rule
- [CSS at-rules](/en-US/docs/Web/CSS/At-rule)

> - [Using feature queries](/en-US/docs/Web/CSS/CSS_Conditional_Rules/Using_Feature_Queries)

Reviewer 2: ```suggestion
- [Using feature queries](/en-US/docs/Web/CSS/CSS_conditional_rules/Using_feature_queries)
```

> - [CSS conditional rules](/en-US/docs/Web/CSS/CSS_Conditional_Rules/) module

Reviewer 2: ```suggestion
- [CSS conditional rules](/en-US/docs/Web/CSS/CSS_conditional_rules/) module
```

> - [CSS conditional rules](/en-US/docs/Web/CSS/CSS_conditional_rules/) module

Reviewer 2: Sorry, I introduced this with my previous suggestion. It's showing up as a flaw
```suggestion
- [CSS conditional rules](/en-US/docs/Web/CSS/CSS_conditional_rules) module
```
